drivers/led_strip/tlc59731.c

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2024 Javad Rahimipetroudi <javad.rahimipetroudi@mind.be>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_tlc59731
+
+/**
+ * @file
+ * @brief LED driver for the TLC59731 LED driver.
+ *
+ * TLC59731 is a 3-Channel, 8-Bit, PWM LED Driver
+ * With Single-Wire Interface (EasySet)
+ *
+ */
+
+#include <zephyr/drivers/led_strip.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(tlc59731, CONFIG_LED_STRIP_LOG_LEVEL);
+
+/* Pulse timing */
+#define DELAY     0x01 /* us */
+#define T_CYCLE_0 0x04 /* us */
+#define T_CYCLE_1 0x01 /* us */
+#define T_H0      (4 * T_CYCLE_0)
+#define T_H1      (8 * T_CYCLE_0)
+/* Threshould levels */
+#define HIGH      0x01
+#define LOW       0x00
+
+/* Write command */
+#define WR 0x3A
+
+struct tlc59731_cfg {
+	struct gpio_dt_spec sdi_gpio;
+	struct led_rgb led_rgb_data;
+};
+
+/*****************************/
+static inline int rgb_pulse(const struct gpio_dt_spec *led_dev)
+{
+	int32_t fret = 0;
+
+	fret = gpio_pin_set_dt(led_dev, HIGH);
+	if (fret != 0) {
+		return fret;
+	}
+
+	fret = gpio_pin_set_dt(led_dev, LOW);
+	if (fret != 0) {
+		return fret;
+	}
+
+	return fret;
+}
+
+static int rgb_eos(void)
+{
+	k_busy_wait(T_H0);
+	return 0;
+}
+
+static int rgb_gslatch(void)
+{
+	k_busy_wait(T_H1);
+
+	return 0;
+}
+
+static int rgb_write_bit(const struct gpio_dt_spec *led_dev, uint8_t data)
+{
+	rgb_pulse(led_dev);
+
+	k_busy_wait(DELAY);
+
+	if (data) {
+		rgb_pulse(led_dev);
+		k_busy_wait(T_CYCLE_1);
+	} else {
+		k_busy_wait(T_CYCLE_0);
+	}
+
+	return 0;
+}
+
+static int rgb_write_data(const struct gpio_dt_spec *led_dev, uint8_t data)
+{
+	int8_t idx = 7;
+
+	while (idx >= 0) {
+		rgb_write_bit(led_dev, data & BIT((idx--)));
+	}
+
+	return 0;
+}
+
+static int rgb_write_led(const struct gpio_dt_spec *led_dev, struct led_rgb *rgb_data)
+{
+	rgb_write_data(led_dev, WR);
+	rgb_write_data(led_dev, rgb_data->r);
+	rgb_write_data(led_dev, rgb_data->g);
+	rgb_write_data(led_dev, rgb_data->b);
+
+	rgb_eos();
+	rgb_gslatch();
```

Comment:
This is wrong: the gslatch should only be done after writing _all_ LEDs, so at the end of the update_rgb loop.

Also, no EOS needs to be done for the last LED (though it's in fact harmless because it's just a delay).